### PR TITLE
SDHC: Add Cadence SDHC driver

### PR DESCRIPTION
**Pull request by roymurlidhar**
- adding driver for cadence sdhc
- adding shell support
- adding driver support for agilex5 platform
- adding agilex5 board support for fat_fs sample application
- adding agilex5 support for shell module

### CODE REVIEW

**Comment by roymurlidhar** [comment]
Hi @danieldegrasse danieldegrasse, Can you review and approve this PR.

File: drivers/sdhc/Kconfig.sdhc_cdns
Comment (danieldegrasse) [inline drivers/sdhc/Kconfig.sdhc_cdns:4]: Why is this definition needed?

File: drivers/sdhc/sdhc_cdns.c
Comment (danieldegrasse) [inline drivers/sdhc/sdhc_cdns.c:82]: You shouldn't need to switch on the opcode here- the `cmd` parameter should already have the `response_type` field set by the SD subystem

File: dts/bindings/sdhc/cdns,sdhc.yaml
Comment (danieldegrasse) [inline dts/bindings/sdhc/cdns,sdhc.yaml:15]: Please add additional documentation for the DT properties you are defining here, at a minimum a `description` value

**Review by danieldegrasse — COMMENTED** [review]
Sorry for the delay getting to this PR- I've added some initial comments

File: drivers/sdhc/sdhc_cdns.c
Comment (roymurlidhar) [reply drivers/sdhc/sdhc_cdns.c:1]: I an fine with either way. Please suggest if we need to do in a single folder under drivers/sdhc/sdhc_cdns.

**Comment by danieldegrasse** [comment]
Hi @roymurlidhar, Zephyr does not allow merge commits. Please take a look at the contribution guidelines for how to rebase your branch on top of the latest `main` branch, particularly section 14: https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow

File: drivers/sdhc/Kconfig.sdhc_cdns
Comment (roymurlidhar) [reply drivers/sdhc/Kconfig.sdhc_cdns:10]: addressed

File: drivers/sdhc/Kconfig.sdhc_cdns
Comment (roymurlidhar) [reply drivers/sdhc/Kconfig.sdhc_cdns]: Removed.

File: drivers/sdhc/sdhc_cdns.c
Comment (roymurlidhar) [reply drivers/sdhc/sdhc_cdns.c]: Corrected.

File: dts/bindings/sdhc/cdns,sdhc.yaml
Comment (roymurlidhar) [reply dts/bindings/sdhc/cdns,sdhc.yaml:17]: Done.

**Comment by roymurlidhar** [comment]
@danieldegrasse  I have addressed all the comments. Please review it.

File: drivers/sdhc/sdhc_cdns.c
Comment (danieldegrasse) [reply drivers/sdhc/sdhc_cdns.c:1]: Sorry, rereading my previous comment it was very unclear. I'd prefer for all related driver files to be in one folder here, like so:
```
drivers/sdhc/sdhc_cdns/sdhc_cdns.c
drivers/sdhc/sdhc_cdns/sdhc_cdns_ll.c
drivers/sdhc/sdhc_cdns/sdhc_cdns_ll.h
```

File: drivers/sdhc/sdhc_cdns.c
Comment (danieldegrasse) [inline drivers/sdhc/sdhc_cdns.c:82]: Two things here:
- these comments will fail CI, Zephyr only allows C89 style `/* */` comments
- Why is this block commented out? It seems like it would be required for the driver to function. Is the implementation of this now in `sdhc_cdns_ll.c`?

**Comment by roymurlidhar** [comment]
@danieldegrasse I am checking other driver files in the community like I2C. In that there is no separate folder for a single IP.
Do we need to create separate folder for this folder drivers/sdhc/sdhc_cdns/"files" ?

**Comment by danieldegrasse** [comment]
> @danieldegrasse I am checking other driver files in the community like I2C. These drivers also consists multiple driver files but in that there is no separate folder for a single IP.
Do we need to create separate folder for this folder drivers/sdhc/sdhc_cdns/"files" ?

This is true, there are a few examples where this has been done before. Given that, I'm fine if you'd prefer to keep these files at the root folder level.

File: drivers/sdhc/sdhc_cdns.c
Comment (roymurlidhar) [reply drivers/sdhc/sdhc_cdns.c:82]: Right now, the Cadence IP6061 driver bare metal functions are abstracted in sdhc_cdns_ll.c files and as of now all these function pointers are pointing to single function since only this IP supported. In future, if any new Cadence IP from the same family is introduced then we will make use of these function pointers to point different sdhc_cdns_ll.c file and by keeping sdhc_cdns.c wrapper file same across both the drivers.

File: samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf
Comment (roymurlidhar) [reply samples/subsys/shell/shell_module/boards/intel_socfpga_agilex5_socdk.conf:31]: The intention here is to provide all the features enabled as part of our development kit to our customers - one shell application to get access to all the hardware features.

File: drivers/sdhc/sdhc_cdns.c
Comment (danieldegrasse) [reply drivers/sdhc/sdhc_cdns.c:82]: Understood. If this abstraction layer will make porting future IP to Zephyr easier than I am ok with it. Just to clarify- is the sdhc_cdns_ll.c a wrapper that exists for other IP not presently in Zephyr? Or is this wrapper being designed specifically for Zephyr support?

Comment (roymurlidhar) [reply drivers/sdhc/sdhc_cdns.c:82]: This is designed specially for zephyr support.

Comment (decsny) [reply drivers/sdhc/sdhc_cdns.c:82]: @roymurlidhar do you expect there to be multiple cadence SDHCs of different IP types/versions on the same platform? because that would be the main justification for function pointers instead of just compiling different definitions for these functions

Comment (decsny) [reply drivers/sdhc/sdhc_cdns.c:82]: also, if this is all zephyr specific/native anyways, what benefit do you get from having another abstraction layer, when you can write just a different zephyr driver with one layer for the different IP

Comment (gdengi) [reply drivers/sdhc/sdhc_cdns.c:82]: @danieldegrasse @decsny 

> also, if this is all zephyr specific/native anyways, what benefit do you get from having another abstraction layer, when you can write just a different zephyr driver with one layer for the different IP
> 
> a lot of the functions you're implementing in the 2nd layer are 1 to 1 with the sdhc api that already exists and you are implementing, for example your reset function implementation of the zephyr api is just calling your 2nd layer reset api and doing nothing else, what is the purpose of this, this is why there is already a first layer api for this, because different devices have different implementations for it.

Hello @decsny,

Agree with your comments. However, our main objective of splitting this driver into 2 layers is:
1st layer of the driver will handle all the Zephyr device driver framework related code
2nd layer will have the bare metal code (Independent of Zephyr or any other OS) for all the functionalities of the IP.

The motive behind this abstraction is: to provide Zephyr driver and as well as bare metal code also to our (Intel Corporation) customers.

Comment (gdengi) [reply drivers/sdhc/sdhc_cdns.c:82]: > Ok. For my own understanding- is this abstraction layer going to enable reuse in the future? IE would a future driver for different IP possibly reuse functions from the sdhc_cdns_ll.c file? I'm trying to understand what the benefit/reasoning is for introducing this additional layer of abstraction

Hello @danieldegrasse,

I quoted our main objective of this abstraction layer to @decsny reply.
Yes, if the changes are minimal, we can re-use the same abstraction layer for different Cadence IP from the same family.

Thank you.

Comment (danieldegrasse) [reply drivers/sdhc/sdhc_cdns.c:82]: @gdengi thank you for the clarification. Provided that this layer can be reused within Zephyr, I am ok with the approach.

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/sdhc/sdhc_cdns.c
Comment (gdengi) [reply drivers/sdhc/sdhc_cdns.c:82]: Thank you @danieldegrasse for your valuable feedbacks and review comments.

Comment (roymurlidhar) [reply drivers/sdhc/sdhc_cdns.c:82]: @decsny @danieldegrasse Is there any feedback or comments. If not can we approve this PR.

**Comment by roymurlidhar** [comment]
@decsny  @danieldegrasse  Is there any feedback or comments. If not can we approve this PR.

**Review by decsny — DISMISSED** [review]

**Comment by carlescufi** [comment]
@roymurlidhar sorry, can you please rebase this PR?

**Review by decsny — DISMISSED** [review]

File: CODEOWNERS
Comment (danieldegrasse) [inline CODEOWNERS:368]: Why are these lines related to `timer` being added to codeowners?

**Comment by danieldegrasse** [comment]
@roymurlidhar please address compliance failures when you get a chance

**Comment by roymurlidhar** [comment]
> @roymurlidhar please address compliance failures when you get a chance

@danieldegrasse @decsny @carlescufi  all checks are passed. Can we approve it and merge?

**Review by decsny — DISMISSED** [review]

File: CODEOWNERS
Comment (danieldegrasse) [reply CODEOWNERS:368]: @roymurlidhar can you please clarify why the line regarding `timers` is being added? I don't see how it relates to the scope of this PR.

**Comment by roymurlidhar** [comment]
> > @roymurlidhar please address compliance failures when you get a chance
> 
> @danieldegrasse @decsny @carlescufi all checks are passed. Can we approve it and merge?

@danieldegrasse  Can you approve this PR and merge it. I think all the comments has been acknowledged.

**Comment by danieldegrasse** [comment]
> @danieldegrasse Can you approve this PR and merge it. I think all the comments has been acknowledged.

>> @roymurlidhar can you please clarify why the line regarding timers is being added? I don't see how it relates to the scope of this PR.

Can you please clarify the above question? I am fine with the contents of this PR besides those changes- I would like to see the non-sdhc changes to the codeowners file removed from this PR or justified.

**Comment by roymurlidhar** [comment]
@danieldegrasse I have addressed your comment for "CODEOWNERS" @decsny   Please approve this PR.

**Review by decsny — DISMISSED** [review]

File: CODEOWNERS
Comment (danieldegrasse) [inline CODEOWNERS:368]: @roymurlidhar Please drop this line from the codeowners file

**Review by danieldegrasse — APPROVED** [review]

**Comment by roymurlidhar** [comment]
@decsny Please approve. @danieldegrasse Please merge it once it got approval.

**Review by decsny — APPROVED** [review]

**Comment by danieldegrasse** [comment]
> @decsny Please approve. @danieldegrasse Please merge it once it got approval.

@roymurlidhar now that #66442 is merged, you'll need to rebase this PR onto the latest revision of `main` to restart CI and get everything passing. Sorry about that!

**Comment by decsny** [comment]
You don't need to revise unless there is a conflict, just rerun ci

**Comment by roymurlidhar** [comment]
Hi @decsny and @danieldegrasse can we merge this?

**Pull request merged**
